File: elements-of-programming-interviews/2023/python/int_as_list_add.py
from typing import Optional

from list_node import ListNode
from test_framework import generic_test

# TODO: Negatives


def add_two_numbers(L1: ListNode, L2: ListNode) -> Optional[ListNode]:
    # NOTE: Only works for nonnegatives
    if not L2 or not L1:
        return L1 if L1 else L2

    head = curr = ListNode(-1, None)
    carry = 0
    while L1 or L2 or carry:
        op1 = L1.data if L1 else 0
        op2 = L2.data if L2 else 0
        digit = (op1 + op2 + carry) % 10
        carry = (op1 + op2 + carry) // 10
        curr.next = ListNode(digit, None)
        curr = curr.next
        L1 = L1.next if L1 else L1
        L2 = L2.next if L2 else L2

    return head.next


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('int_as_list_add.py',
                                       'int_as_list_add.tsv', add_two_numbers))

File: elements-of-programming-interviews/2023/cpp/int_as_list_add.cc
#include "list_node.h"
#include "test_framework/generic_test.h"

/*
  - keep place, iterate through, (digit + digit) * place
  - need carry if adding node by node.
  - negatives? can handle by reading both to ints first then adding
  - no empty list
*/

shared_ptr<ListNode<int>> AddTwoNumbers(shared_ptr<ListNode<int>> L1,
                                        shared_ptr<ListNode<int>> L2) {
  int carry = 0;
  int total = 0;
  shared_ptr<ListNode<int>> current = make_shared<ListNode<int>>();
  shared_ptr<ListNode<int>> head = current;
  while (L1 || L2 || carry) {
    int val1 = (L1 != nullptr) ? L1->data : 0;
    int val2 = (L2 != nullptr) ? L2->data : 0;
    L1 = (L1 == nullptr) ? nullptr : L1->next;
    L2 = (L2 == nullptr) ? nullptr : L2->next;

    // Assuming no negatives.
    total = (val1 + val2 + carry) % 10;
    carry = (val1 + val2 + carry >= 10) ? 1 : 0;

    current->next = make_shared<ListNode<int>>();
    current->next->data = total;
    current = current->next;
  }
  return head->next;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"L1", "L2"};
  return GenericTestMain(args, "int_as_list_add.cc", "int_as_list_add.tsv",
                         &AddTwoNumbers, DefaultComparator{}, param_names);
}
